scalyr_agent/log_processing.py

```diff
@@ -1866,16 +1889,98 @@ def __apply_redaction_rule(self, line, redaction_rule):
         @return: A sequence of two elements, the line with the redaction applied (if any) and True or False
             indicating if a redaction was applied.
         """
+
+        def __replace_groups_with_hashed_content(re_ex, replacement_ex, line):
```

Comment:
I think you can simplify this code a fair bit.  Here's the approach I would take:

Suppose you have regex string `password=(.*)&` and replacement `password=\\H1`.

First, only attempt to apply the hash substitution if there is a match for `\\H\d` in the replacement string.

If that is true, then do the following:
* Perform the search, so that you get values for all of the groups.
* for 1 to the number of groups, do the following:
* replace `\\HN` in the replacement string with the hash of group(N)
* end for loop
* Then perform the replacement on the original string, using the new replacement string that has the actual digests built into it.

I think that works as long as we aren't using "global" replacement, which I don't think we are?  You could check that.  If we are, then we just have to iterate overall of the matches in the string and do them one at a time like how I specified above.

What do you think?

Reply:
it does make sense in retrospect. I'll try it out. thank you for your feedback.

Reply:
@czerwingithub  i tried using this approach and it's elegant for single matches, but I couldn't figure out how to make it cleaner than what it already is for multiple matches in a line. If there are multiple matches 
eg: 

`replacement_string`: `secret\1=\H2`
`regular expression`: `secret(.*?)=([a-z]+\s?)`
`log line`: `sometext.... secretoption=czerwin moretextsecretbar=xxx andsecret123=saurabh`

in this case, we will have to iterate over all the matches i.e. `secretoption=czerwin ` `secretbar=xxx ` and `secret123=saurabh` and replace the replacement string inside the loop and make sure the final replaced string needs to be partially replaced for each matched group. This calls for bookeeping on - 
1) making sure the replacement string is replenished after each match occurence
2) the match count is incremented and accounted for
3) the main log line should only be partially matched and we should account for the location until where the string was replaced as part of the first match. This is because replacing the first match with its hash will convert
```sometext.... secretoption=czerwin moretextsecretbar=xxx andsecret123=saurabh``` -> ```sometext.... secretoption=<hash of czerwin> moretextsecretbar=xxx andsecret123=saurabh``` and doing it again will make the first match again susceptible to being a candidate for replacement i.e.
```sometext.... secretoption=<hash of <hash of czerwin>> moretextsecretbar=xxx andsecret123=saurabh```
unless we track the last replaced line index.

This caused my code to be more unclear than what it was before. Let me know what you think.

Reply:
I think you can essentially add an outer loop to applying my suggested replacement trick one replacement at a time.

Use `re.finditer` to walk over each match in a given line.  That returns a `MatchObject` for each match.  You can use its `start` and `end` methods to determine the indices into the original line where the match appears.

You can use this and I think it works out to be fairly clean.

    myindex = 0
    output_result = ""
    for x in the_regex.finditer(the_line):
        output_result = output_result + the_line[my_index:x.start(0)]
        my_index = x.end(0)
        output_result += perform_replacement_for_single(x.group(0), replacement_string)

Does that make sense?  I'm building up the overall string manually but always first adding the portion of the line that did not match replacement pattern (since our last match).  Then, I also add in the text that did match the replacement pattern with the replacements performed by just performing the replacement on x.group(0) [which is the substring that matched the pattern].  Then I iterate from there.